Returning compiling errors

It's really difficult to get the component included in the ESP32-S3.
I tried both the Command Line Interface for ESPhome and docker, but did not get a binary I could flash. 
Eventually decided to install ESPHome inside Home Assistant. 

My YAML file contains this passage:

```
# Externe component van GitHub (Daikin S21)
external_components:
  - source: github://asund/esphome-daikin-s21@main
    components: [ daikin_s21 ]
```

But I keep getting the error below.

In file included from **src/esphome/components/daikin_s21/climate/daikin_s21_climate.cpp:5:**
src/esphome/components/daikin_s21/climate/daikin_s21_climate.h:4:10: fatal error: esphome/components/sensor/sensor.h: No such file or directory
    4 | #include **"esphome/components/sensor/sensor.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.**

Is there a workaround?

--- a/README.md
+++ b/README.md
@@ -5,7 +5,7 @@ protocol available over S21 and related ports.
 
 A big thanks to:
 * [revk](https://github.com/revk) for work on the fantastic
-  [Faikin](https://github.com/revk/ESP32-Faikin) project, which was the primary
+  [Faikout](https://github.com/revk/ESP32-Faikout) (née Faikin) project, which was the primary
   inspiration and guide for building this ESPHome component. In addition, the
   very active and resourceful community that project has fostered that has
   decoded much of the protocol.
@@ -92,7 +92,7 @@ Not all units support these.
 
 ### Text Sensor
 
-If you've read the Faikin wiki you'll see many more queries available than
+If you've read the Faikout wiki you'll see many more queries available than
 what this project supports. I've added a text_sensor component to read these
 raw values out for debugging and protocol decoding use, without having to add
 a dedicated sensor. Normally you wouldn't need to use this, but if a value
@@ -124,7 +124,7 @@ this writing, independent UART pin inversion control also wasn't possible.
 
 ## Hardware
 
-Please see the Faikin [wiring](https://github.com/revk/ESP32-Faikin/wiki/Wiring)
+Please see the Faikout [wiring](https://github.com/revk/ESP32-Faikout/wiki/Wiring)
 page for detailed documentation including pinouts, alternate connectors with
 images. The below is just a quick reference overview.
 
@@ -177,7 +177,7 @@ Contacts: JST `SXA-001T-P0.6`
 
 ### PCB Option 1
 
-joshbenner uses the board designed by revk available [here](https://github.com/revk/ESP32-Faikin/tree/main/PCB/Faikin).
+joshbenner uses the board designed by revk available [here](https://github.com/revk/ESP32-Faikout/tree/main/PCB/Faikout).
 Note that revk's design includes a FET that inverts the logic levels on the
 ESP's RX pin, and on newer revisions the TX pin as well. When interfacing
 through a FET the RX line should be configured with a pullup. This handling
@@ -206,7 +206,7 @@ ways:
   other sensor values.
 
 Please consider documenting characteristics of your unit in the appropriate
-place in the Faikin wiki. I don't want to make an inferior copy of this
+place in the Faikout wiki. I don't want to make an inferior copy of this
 information if possible. Please don't report issues with this component to them
 as we don't share a codebase. Some of the readout values here are not
 byteswapped if we don't otherwise use them so unknown static fields may appear
@@ -246,6 +246,9 @@ esp32:
 
 logger:
   baud_rate: 0  # Disable UART logger if using UART0 (pins 1,3)
+
+preferences:
+  flash_write_interval: 72h
 
 external_components:
   - source: github://asund/esphome-daikin-s21@main
@@ -387,7 +390,7 @@ binary_sensor:
 ```
 
 Here is an example of how daikin_s21 can be used with inverted UART pins
-and an RX pullup (newer Faikin PCB rev):
+and an RX pullup (newer Faikout PCB rev):
 
 ```yaml
 uart:

--- a/components/daikin_s21/binary_sensor/daikin_s21_binary_sensor.h
+++ b/components/daikin_s21/binary_sensor/daikin_s21_binary_sensor.h
@@ -4,6 +4,7 @@
 #include "esphome/core/component.h"
 #include "esphome/core/helpers.h"
 #include "../daikin_s21_types.h"
+#include "../s21.h"
 
 namespace esphome::daikin_s21 {
 
@@ -16,27 +17,34 @@ class DaikinS21BinarySensor : public Component,
 
   void set_powerful_sensor(binary_sensor::BinarySensor *sensor) {
     this->powerful_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutUnitStateBits);
   }
   void set_defrost_sensor(binary_sensor::BinarySensor *sensor) {
     this->defrost_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutUnitStateBits);
   }
   void set_active_sensor(binary_sensor::BinarySensor *sensor) {
     this->active_sensor_ = sensor;
   }
   void set_online_sensor(binary_sensor::BinarySensor *sensor) {
     this->online_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutUnitStateBits);
   }
   void set_valve_sensor(binary_sensor::BinarySensor *sensor) {
     this->valve_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutSystemStateBits);
   }
   void set_short_cycle_sensor(binary_sensor::BinarySensor *sensor) {
     this->short_cycle_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutSystemStateBits);
   }
   void set_system_defrost_sensor(binary_sensor::BinarySensor *sensor) {
     this->system_defrost_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutSystemStateBits);
   }
-    void set_multizone_conflict_sensor(binary_sensor::BinarySensor *sensor) {
+  void set_multizone_conflict_sensor(binary_sensor::BinarySensor *sensor) {
     this->multizone_conflict_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutSystemStateBits);
   }
 
  protected:

--- a/components/daikin_s21/climate/__init__.py
+++ b/components/daikin_s21/climate/__init__.py
@@ -18,6 +18,8 @@ from .. import (
     S21_PARENT_SCHEMA,
 )
 
+AUTO_LOAD = ["sensor"]
+
 DaikinS21Climate = daikin_s21_ns.class_(
     "DaikinS21Climate", climate.Climate, cg.PollingComponent
 )
@@ -41,7 +43,6 @@ CONF_MAX_COOL_TEMPERATURE = "max_cool_temperature"
 CONF_MIN_COOL_TEMPERATURE = "min_cool_temperature"
 CONF_MAX_HEAT_TEMPERATURE = "max_heat_temperature"
 CONF_MIN_HEAT_TEMPERATURE = "min_heat_temperature"
-
 
 CONFIG_SCHEMA = (
     climate.climate_schema(DaikinS21Climate)

--- a/components/daikin_s21/climate/daikin_s21_climate.cpp
+++ b/components/daikin_s21/climate/daikin_s21_climate.cpp
@@ -137,9 +137,7 @@ void DaikinS21Climate::loop() {
         this->check_setpoint = false;
         if (DaikinC10::diff(reported.setpoint, this->calc_s21_setpoint()) >= SETPOINT_STEP) { // in setpoint mode, calculated won't be invalid
           if (this->use_temperature_sensor()) {
-            ESP_LOGD(TAG, "Temperature from external sensor: %.1f %s (%.1f °C)  Offset: %.1f",
-                this->temperature_sensor_->get_state(),
-                this->temperature_sensor_->get_unit_of_measurement_ref().c_str(),
+            ESP_LOGD(TAG, "Temperature from external sensor: %.1f Offset: %.1f",
                 current_temperature,
                 this->get_parent()->get_temp_inside() - this->temperature_sensor_degc());
           }
@@ -235,6 +233,7 @@ void DaikinS21Climate::set_supported_modes(std::set<climate::ClimateMode> modes)
  */
 void DaikinS21Climate::set_supported_presets(std::set<climate::ClimatePreset> presets) {
   this->traits_.set_supported_presets(presets);
+  this->get_parent()->request_readout(DaikinS21::ReadoutPresets);
 }
 
 /**
@@ -322,7 +321,7 @@ DaikinC10 DaikinS21Climate::calc_s21_setpoint() {
 }
 
 void DaikinS21Climate::save_setpoint(const DaikinC10 value) {
-  // Only save if value is different from what's already saved.
+  // Only save if value is different from what's already saved. Some platforms don't support this internally.
   if (value != this->load_setpoint()) {
     const int16_t save_val = static_cast<int16_t>(value);
     switch (this->mode) {

--- a/components/daikin_s21/s21.cpp
+++ b/components/daikin_s21/s21.cpp
@@ -1,6 +1,7 @@
 #include <cinttypes>
 #include <numeric>
 #include <ranges>
+#include "daikin_s21_queries.h"
 #include "s21.h"
 #include "utils.h"
 
@@ -106,31 +107,25 @@ const char * active_source_to_string(const ActiveSource source) {
   }
 }
 
+/**
+ * Convert reversed ASCII number to an integer
+ *
+ * @param bytes ASCII bytes of format <ones><tens><hundreds[><neg/pos>,<thousands>]
+ * @return int16_t integer representation of string
+ */
 int16_t bytes_to_num(std::span<const uint8_t> bytes) {
-  // <ones><tens><hundreds[><neg/pos>,<thousands>]
-  int16_t val = 0;
-  val = bytes[0] - '0';
-  val += (bytes[1] - '0') * 10;
-  val += (bytes[2] - '0') * 100;
-  if (bytes.size() > 3) {
-    if (bytes[3] == '-') {
-      val *= -1;
-    } else if (std::isdigit(bytes[3])) {
-      val += (bytes[3] - '0') * 1000;
-    }
-  }
-  return val;
+  std::array<char, 4+1> buffer{};
+  std::ranges::reverse_copy(bytes, buffer.begin());
+  return std::atoi(buffer.data());
 }
 
 void DaikinS21::setup() {
   // populate initial messages to poll
-  // clang-format off
   this->queries = {
       // Protocol version detect
       {StateQuery::OldProtocol, &DaikinS21::handle_nop, true},
       {StateQuery::NewProtocol, &DaikinS21::handle_nop, true},
   };
-  // clang-format on
   this->failed_queries = {};
   this->static_queries = {};
   this->protocol_version = ProtocolUndetected;
@@ -297,13 +292,9 @@ void DaikinS21::refine_queries() {
     }
     if (this->protocol_version >= ProtocolVersion(2)) {
       this->queries.insert(this->queries.end(), {
-        {StateQuery::SpecialModes, &DaikinS21::handle_state_special_modes},
-        {StateQuery::DemandAndEcono, &DaikinS21::handle_state_demand_and_econo},
         // {StateQuery::FB, &DaikinS21::handle_nop}, // unknown
         {StateQuery::ModelCode, &DaikinS21::handle_state_model_code_v2, true},
-        {StateQuery::IRCounter, &DaikinS21::handle_state_ir_counter},
         {StateQuery::V2OptionalFeatures, &DaikinS21::handle_nop, true},
-        {StateQuery::PowerConsumption, &DaikinS21::handle_state_power_consumption},
         // {StateQuery::ITELC, &DaikinS21::handle_nop},  // unknown, daikin intelligent touch controller?
         // {StateQuery::FP, &DaikinS21::handle_nop}, // unknown
         // {StateQuery::FQ, &DaikinS21::handle_nop}, // unknown
@@ -313,6 +304,18 @@ void DaikinS21::refine_queries() {
         // {StateQuery::NewProtocol, &DaikinS21::handle_nop, true},  // already added in initial detection
         {MiscQuery::SoftwareVersion, &DaikinS21::handle_misc_software_version, true},
       });
+      if (this->readout_requests[ReadoutPresets]) {
+        this->queries.insert(this->queries.end(), {
+          {StateQuery::SpecialModes, &DaikinS21::handle_state_special_modes},
+          {StateQuery::DemandAndEcono, &DaikinS21::handle_state_demand_and_econo},
+        });
+      }
+      if (this->readout_requests[ReadoutIRCounter]) {
+        this->queries.emplace_back(StateQuery::IRCounter, &DaikinS21::handle_state_ir_counter);
+      }
+      if (this->readout_requests[ReadoutPowerConsumption]) {
+        this->queries.emplace_back(StateQuery::PowerConsumption, &DaikinS21::handle_state_power_consumption);
+      }
     }
     // todo >= ProtocolVersion(3,0)
     // common sensors
@@ -325,16 +328,11 @@ void DaikinS21::refine_queries() {
       // {EnvironmentQuery::SwingMode, &DaikinS21::handle_env_swing_mode},  // redundant
       // {EnvironmentQuery::FanMode, &DaikinS21::handle_env_fan_mode},  // added in optional feature detection
       {EnvironmentQuery::InsideTemperature, &DaikinS21::handle_env_inside_temperature},
-      {EnvironmentQuery::LiquidTemperature, &DaikinS21::handle_env_liquid_temperature},
       // {EnvironmentQuery::FanSpeedSetpoint, &DaikinS21::handle_env_fan_speed_setpoint},  // not supported yet, can translate DaikinFanMode to RPM
       // {EnvironmentQuery::FanSpeed, &DaikinS21::handle_env_fan_speed}, // added in optional feature detection
       // {EnvironmentQuery::LouvreAngleSetpoint, &DaikinS21::handle_env_vertical_swing_angle_setpoint},  // not supported yet
       // {EnvironmentQuery::VerticalSwingAngle, &DaikinS21::handle_env_vertical_swing_angle}, // added in optional feature detection
       // {EnvironmentQuery::RW, &DaikinS21::handle_nop},  // unknown, "00" for me
-      {EnvironmentQuery::TargetTemperature, &DaikinS21::handle_env_target_temperature},
-      {EnvironmentQuery::OutsideTemperature, &DaikinS21::handle_env_outside_temperature},
-      {EnvironmentQuery::IndoorFrequencyCommandSignal, &DaikinS21::handle_env_indoor_frequency_command_signal},
-      {EnvironmentQuery::CompressorFrequency, &DaikinS21::handle_env_compressor_frequency},
       // {EnvironmentQuery::IndoorHumidity, &DaikinS21::handle_env_indoor_humidity},  // added in optional feature detection
       // {EnvironmentQuery::CompressorOnOff}, // redundant
       // {EnvironmentQuery::UnitState, &DaikinS21::handle_env_unit_state}, // added in model detection
@@ -342,6 +340,21 @@ void DaikinS21::refine_queries() {
       // {EnvironmentQuery::Rz52, &DaikinS21::handle_nop},  // unknown, "40"for me
       // {EnvironmentQuery::Rz72, &DaikinS21::handle_nop},  // unknown, "23" for me
     });
+    if (this->readout_requests[ReadoutTemperatureCoil]) {
+      this->queries.emplace_back(EnvironmentQuery::LiquidTemperature, &DaikinS21::handle_env_liquid_temperature);
+    }
+    if (this->readout_requests[ReadoutTemperatureTarget]) {
+      this->queries.emplace_back(EnvironmentQuery::TargetTemperature, &DaikinS21::handle_env_target_temperature);
+    }
+    if (this->readout_requests[ReadoutTemperatureOutside]) {
+      this->queries.emplace_back(EnvironmentQuery::OutsideTemperature, &DaikinS21::handle_env_outside_temperature);
+    }
+    if (this->readout_requests[ReadoutDemand]) {
+      this->queries.emplace_back(EnvironmentQuery::IndoorFrequencyCommandSignal, &DaikinS21::handle_env_indoor_frequency_command_signal);
+    }
+    if (this->readout_requests[ReadoutCompressorFrequency]) {
+      this->queries.emplace_back(EnvironmentQuery::CompressorFrequency, &DaikinS21::handle_env_compressor_frequency);
+    }
   }
 
   // Protocol detected and initial queries scheduled
@@ -358,20 +371,21 @@ void DaikinS21::refine_queries() {
     // handle results
     if (this->ready[ReadyOptionalFeatures]) {
       // v2 GK info gates base G2 support if present
-      if (v2_features.nak || (v2_features.ack && (v2_features.value)[2] & 0b00000100)) {
+      if (v2_features.nak || (v2_features.ack && (v2_features.value[2] & 0b00000100))) {
         // swing
         this->support.swing = (features.ack && (features.value[0] & 0b0100));
         if (this->support.swing) {
           this->support.horizontal_swing = (features.ack && (features.value[0] & 0b1000));
-          this->queries.emplace_back(EnvironmentQuery::VerticalSwingAngle, &DaikinS21::handle_env_vertical_swing_angle);
+          if (this->readout_requests[ReadoutSwingAngle]) {
+            this->queries.emplace_back(EnvironmentQuery::VerticalSwingAngle, &DaikinS21::handle_env_vertical_swing_angle);
+          }
         }
-
         // fan
         this->support.fan = true;
-        this->queries.insert(this->queries.end(), {
-          {EnvironmentQuery::FanMode, &DaikinS21::handle_env_fan_mode},
-          {EnvironmentQuery::FanSpeed, &DaikinS21::handle_env_fan_speed},
-        });
+        this->queries.emplace_back(EnvironmentQuery::FanMode, &DaikinS21::handle_env_fan_mode);
+        if (this->readout_requests[ReadoutFanSpeed]) {
+          this->queries.emplace_back(EnvironmentQuery::FanSpeed, &DaikinS21::handle_env_fan_speed);
+        }
       }
 
       // humidity
@@ -379,7 +393,9 @@ void DaikinS21::refine_queries() {
         this->support.humidity = (features.ack && (features.value[0] & 0b0010));
         if (this->support.humidity) {
           // unknown if this is a dehumidify mode or humidity sensor. potentially hides the sensor if it's just the mode. my unit supports neither over S21 interface.
-          this->queries.emplace_back(EnvironmentQuery::IndoorHumidity, &DaikinS21::handle_env_indoor_humidity);
+          if (this->readout_requests[ReadoutHumidity]) {
+            this->queries.emplace_back(EnvironmentQuery::IndoorHumidity, &DaikinS21::handle_env_indoor_humidity);
+          }
         }
       }
 
@@ -410,7 +426,10 @@ void DaikinS21::refine_queries() {
     const auto outside = this->get_query_result(EnvironmentQuery::OutsideTemperature);
     const auto humidity = this->get_query_result(EnvironmentQuery::IndoorHumidity);
     // done if all queries have been resolved
-    this->ready[ReadySensorReadout] = (fan_mode && inside && outside && humidity);
+    this->ready[ReadySensorReadout] = fan_mode &&
+                                      inside &&
+                                      ((this->readout_requests[ReadoutTemperatureOutside] == false) || outside) &&
+                                      ((this->readout_requests[ReadoutHumidity] == false) || humidity);
     // handle results
     if (this->ready[ReadySensorReadout]) {
       this->support.fan_mode_query = fan_mode.ack;
@@ -418,7 +437,9 @@ void DaikinS21::refine_queries() {
       this->support.outside_temperature_query = outside.ack;
       this->support.humidity_query = humidity.ack;
       // enable coarse fallback query if any sensor query failed
-      const bool alt = (inside.nak || outside.nak || (this->support.humidity && humidity.nak)); // only enable for humidity's sake if declared to be supported
+      const bool alt = inside.nak ||
+                       (this->readout_requests[ReadoutTemperatureOutside] && outside.nak) ||
+                       (this->readout_requests[ReadoutHumidity] && this->support.humidity && humidity.nak); // only enable for humidity's sake if declared to be supported
       if (alt) {
         this->queries.emplace_back(StateQuery::InsideOutsideTemperatures, &DaikinS21::handle_state_inside_outside_temperature);
       }
@@ -446,10 +467,12 @@ void DaikinS21::refine_queries() {
           break;
       }
       if (this->support.unit_system_state_queries) {
-        this->queries.insert(this->queries.end(), {
-          {EnvironmentQuery::UnitState, &DaikinS21::handle_env_unit_state},
-          {EnvironmentQuery::SystemState, &DaikinS21::handle_env_system_state},
-        });
+        if (this->readout_requests[ReadoutUnitStateBits]) {
+          this->queries.emplace_back(EnvironmentQuery::UnitState, &DaikinS21::handle_env_unit_state);
+        }
+        if (this->readout_requests[ReadoutSystemStateBits]) {
+          this->queries.emplace_back(EnvironmentQuery::SystemState, &DaikinS21::handle_env_system_state);
+        }
       }
       // no v2 handling required yet
       ESP_LOGD(TAG, "Model %04" PRIX16 " %04" PRIX16 " detected", this->modelV0, this->modelV2);
@@ -464,15 +487,15 @@ void DaikinS21::refine_queries() {
         this->support.active_source = ActiveSource::CompressorOnOff;
       } else {
         const auto unit_state = this->get_query_result(EnvironmentQuery::UnitState);
-        if (unit_state) {
-          if (unit_state.ack) {
-            this->support.active_source = ActiveSource::UnitState;
-          }
-          if (unit_state.nak) {
-            this->support.unit_system_state_queries = false;  // unit state is assumed to be supported, if it isn't then correct the record
-            this->support.active_source = ActiveSource::Unsupported;
-            this->current.active = true;  // always active, reported action will follow unit
-          }
+        if ((unit_state == QueryNotScheduled) && this->support.unit_system_state_queries) {
+          this->readout_requests.set(ReadoutUnitStateBits);
+          this->queries.emplace_back(EnvironmentQuery::UnitState, &DaikinS21::handle_env_unit_state);
+        } else if (unit_state.ack) {
+          this->support.active_source = ActiveSource::UnitState;
+        } else if (unit_state.nak) {
+          this->support.unit_system_state_queries = false;  // unit state is assumed to be supported, if it isn't then correct the record
+          this->support.active_source = ActiveSource::Unsupported;
+          this->current.active = true;  // always active, reported action will follow unit
         }
       }
     }
@@ -481,6 +504,19 @@ void DaikinS21::refine_queries() {
     if (this->ready[ReadyActiveSource]) {
       ESP_LOGD(TAG, "Active source is %s", active_source_to_string(this->support.active_source));
     }
+  }
+
+  // Select the source of the powerful flag. Depends on the model having been detected.
+  if (this->ready[ReadyModelDetection] && (this->ready[ReadyPowerfulSource] == false)) {
+    if (this->readout_requests[ReadoutPresets] &&
+        (this->get_query_result(StateQuery::SpecialModes) == QueryNotScheduled) &&
+        this->support.unit_system_state_queries &&
+        (this->get_query_result(EnvironmentQuery::UnitState) == QueryNotScheduled)) {
+      this->readout_requests.set(ReadoutUnitStateBits);
+      this->queries.emplace_back(EnvironmentQuery::UnitState, &DaikinS21::handle_env_unit_state);
+      ESP_LOGD(TAG, "Alternate powerful source enabled");
+    }
+    this->ready[ReadyPowerfulSource] = true;
   }
 
   // Finally, schedule any user specified debug queries
@@ -701,7 +737,7 @@ void DaikinS21::handle_env_swing_mode(std::span<uint8_t> &payload) {
   this->current.climate.swing = daikin_to_climate_swing_mode(payload[0]);
 }
 
-/** Better info than StateQuery::Basic (reports quiet) */
+/** Better info than StateQuery::Basic (reports silent) */
 void DaikinS21::handle_env_fan_mode(std::span<uint8_t> &payload) {
   this->current.climate.fan = static_cast<daikin_s21::DaikinFanMode>(payload[0]);
 }
@@ -779,7 +815,7 @@ void DaikinS21::handle_misc_software_version(std::span<uint8_t> &payload) {
 }
 
 /**
- * Determine the protocol version according to Faikin documentation.
+ * Determine the protocol version according to Faikout documentation.
  *
  * @note This is mostly untested, I own only one Daikin system.
  *

--- a/components/daikin_s21/s21.h
+++ b/components/daikin_s21/s21.h
@@ -28,6 +28,30 @@ class DaikinS21 : public PollingComponent {
   // external command action
   void set_climate_settings(const DaikinClimateSettings &settings);
 
+  enum ReadoutRequest {
+    // binary sensor
+    ReadoutUnitStateBits,
+    ReadoutSystemStateBits,
+    // climate
+    ReadoutPresets,
+    // sensor
+    ReadoutTemperatureTarget,
+    ReadoutTemperatureOutside,
+    ReadoutTemperatureCoil,
+    ReadoutFanSpeed,
+    ReadoutSwingAngle,
+    ReadoutCompressorFrequency,
+    ReadoutHumidity,
+    ReadoutDemand,
+    ReadoutIRCounter,
+    ReadoutPowerConsumption,
+    // just for bitset sizing
+    ReadoutCount,
+  };
+  void request_readout(const ReadoutRequest request) {
+    this->readout_requests.set(request);
+  }
+
   std::vector<std::string_view> debug_queries{};
 
   // callbacks called when a query cycle is complete
@@ -35,7 +59,7 @@ class DaikinS21 : public PollingComponent {
 
   // value accessors
   bool is_ready() { return this->ready.all(); }
-  const DaikinClimateSettings& get_climate_settings() { return this->current.climate; };
+  const DaikinClimateSettings& get_climate_settings() { return this->current.climate; }
   auto get_climate_mode() { return this->current.climate.mode; }
   auto get_climate_action() { return this->current.action; }
   auto get_temp_setpoint() { return this->current.climate.setpoint; }
@@ -72,6 +96,7 @@ class DaikinS21 : public PollingComponent {
     ReadySensorReadout,
     ReadyModelDetection,
     ReadyActiveSource,
+    ReadyPowerfulSource,
     ReadyCount, // just for bitset sizing
   };
   std::bitset<ReadyCount> ready{};
@@ -86,6 +111,7 @@ class DaikinS21 : public PollingComponent {
   bool comms_detected() const { return this->protocol_version != ProtocolUndetected; }
   bool cycle_triggered{};
   bool cycle_active{};
+  std::bitset<ReadoutCount> readout_requests{};
   std::vector<DaikinQueryState> queries{};
   std::size_t query_index{};
   auto current_query() { return this->queries.begin() + this->query_index; }

--- a/components/daikin_s21/sensor/daikin_s21_sensor.h
+++ b/components/daikin_s21/sensor/daikin_s21_sensor.h
@@ -4,6 +4,7 @@
 #include "esphome/core/component.h"
 #include "esphome/core/helpers.h"
 #include "../daikin_s21_types.h"
+#include "../s21.h"
 
 namespace esphome::daikin_s21 {
 
@@ -22,33 +23,43 @@ class DaikinS21Sensor : public PollingComponent,
   }
   void set_temp_target_sensor(sensor::Sensor *sensor) {
     this->temp_target_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutTemperatureTarget);
   }
   void set_temp_outside_sensor(sensor::Sensor *sensor) {
     this->temp_outside_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutTemperatureOutside);
   }
   void set_temp_coil_sensor(sensor::Sensor *sensor) {
     this->temp_coil_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutTemperatureCoil);
   }
   void set_fan_speed_sensor(sensor::Sensor *sensor) {
     this->fan_speed_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutFanSpeed);
   }
   void set_swing_vertical_angle_sensor(sensor::Sensor *sensor) {
     this->swing_vertical_angle_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutSwingAngle);
   }
   void set_compressor_frequency_sensor(sensor::Sensor *sensor) {
     this->compressor_frequency_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutCompressorFrequency);
   }
   void set_humidity_sensor(sensor::Sensor *sensor) {
     this->humidity_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutHumidity);
   }
   void set_demand_sensor(sensor::Sensor *sensor) {
     this->demand_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutDemand);
   }
   void set_ir_counter_sensor(sensor::Sensor *sensor) {
     this->ir_counter_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutIRCounter);
   }
   void set_power_consumption_sensor(sensor::Sensor *sensor) {
     this->power_consumption_sensor_ = sensor;
+    this->get_parent()->request_readout(DaikinS21::ReadoutPowerConsumption);
   }
 
  protected:

